File: game_of_life/gol.rb

# 1 = alive
# 0 = dead
class Game
  def initialize
    @grid = make_array
  end

  def make_array
    grid = []
    10.times do |row|
      col = []
      10.times do
        col << rand(0..1)
      end
      grid << col
    end 
    return grid
  end

  def execute
    while true
      @next_gen = make_array
      system("clear")
      render_grid
      update_grid
      @grid = @next_gen
      wait_for_input = gets.chomp
    end
  end

  def render_grid
    @grid.each do |row|
      row.each do |cell|
        print cell.to_s + " "
      end
      print "\n"
    end
  end

  def update_grid
    @grid.each_with_index do |row, row_index|
      row.each_with_index do |cell, cell_index|
        alive_neigbours = check_neighbours(row_index, cell_index)
        if alive_neigbours < 2
          @next_gen[row_index][cell_index] = 0
        elsif alive_neigbours > 3
          @next_gen[row_index][cell_index] = 0
        elsif alive_neigbours == 3
          @next_gen[row_index][cell_index] = 1
        else
          @next_gen[row_index][cell_index] == @grid[row_index][cell_index]
        end
      end
    end
  end

  def check_neighbours(y, x)
    neighbours = 0
    x_buffer = -1
    y_buffer = -1
    9.times do
      if cell_exists?(y + y_buffer, x + x_buffer)
        neighbours += @grid[y + y_buffer][x + x_buffer]
      end
      x_buffer += 1
      if x_buffer == 2
        x_buffer = -1
        y_buffer += 1
      end
    end
    neighbours -= @grid[y][x]
    return neighbours
  end

  def cell_exists?(y,x)
    if (y <= @grid.length - 1) && (y >= 0)
      if (x <= @grid[0].length - 1) && (x >= 0)
        return true
      end
    end
  end
  
end

game = Game.new

game.execute

File: game_of_life/javascript/gol.js

class Game {
  constructor(){
    this.grid = this.makeArray()
  }

  makeArray(){
    let grid = new Array(10)
    for(let i = 0; i < grid.length; i++) {
      var col = []
      for(let j = 0; j < 10; j++){
          col.push(Math.floor(Math.random() * Math.floor(2)))
      } 
      grid[i] = col
    }
    grid = grid
    return grid
  }

  startGame(){
    let gen = 0
    while(gen < 5){
      console.clear()
      let next_gen = this.makeArray()
      console.table(this.grid)
      next_gen = this.update_grid(next_gen)
      //  console.table(next_gen)
      this.grid = next_gen
      alert("Next gen") // This is not the best solution, but it just for basic visuals. A better solution would be using a canvas.
      gen++
    }
  }
  
  update_grid(next_gen){
    for(let row_index = 0; row_index < this.grid.length; row_index++){
      let row = this.grid[row_index]

      for(let cell_index = 0; cell_index < row.length; cell_index++){
        let alive_neighbours = this.check_neighbours(row_index, cell_index)
        // console.log(`Alive ${alive_neighbours} for: ${row_index}, ${cell_index}`)
        if(alive_neighbours < 2){
          next_gen[row_index][cell_index] = 0
        }else if(alive_neighbours > 3){
          next_gen[row_index][cell_index] = 0
        }else if(alive_neighbours === 3){
          next_gen[row_index][cell_index] = 1
        }else{
          next_gen[row_index][cell_index] = this.grid[row_index][cell_index]
        }
      }
    }
    return next_gen
  }

  check_neighbours(y, x){
    let neighbours = 0
    let x_buffer = -1
    let y_buffer = -1
    
    for(let times = 0; times < 9; times++){
      if(this.cellExists(y + y_buffer, x + x_buffer)){
        // console.log(`${y + y_buffer},${x + x_buffer} = ${this.grid[y + y_buffer][x + x_buffer]}` )
        neighbours += this.grid[y + y_buffer][x + x_buffer]
      }
      x_buffer += 1
      if(x_buffer === 2){
        x_buffer = -1
        y_buffer += 1
      }
    }
    neighbours -= this.grid[y][x]
    return neighbours
  }
  
  cellExists(y, x){
    if(y <= this.grid.length - 1 && y >= 0){
      if(x <= this.grid[0].length - 1 && x >= 0){
        return true
      }
    }
  }
}

let mygame = new Game
mygame.startGame()